enterprise/coderd/scim.go

```diff
@@ -0,0 +1,194 @@
+}
+
+// getUsers intentionally always returns no users. This is done to always force
+// Okta to try and create each user individually, this way we don't need to
+// implement fetching users twice.
+//
+//nolint:revive
+func (api *API) scimGetUsers(rw http.ResponseWriter, r *http.Request) {
+	if !api.scimVerifyAuthHeader(r) {
+		_ = handlerutil.WriteError(rw, spec.Error{Status: http.StatusUnauthorized, Type: "invalidAuthorization"})
+		return
+	}
+
+	_ = handlerutil.WriteSearchResultToResponse(rw, &service.QueryResponse{
+		TotalResults: 0,
+		StartIndex:   1,
+		ItemsPerPage: 0,
+		Resources:    []scimjson.Serializable{},
+	})
+}
+
+// scimGetUser intentionally always returns an error saying the user wasn't found.
+// This is done to always force Okta to try and create the user, this way we
+// don't need to implement fetching users twice.
+//
+//nolint:revive
+func (api *API) scimGetUser(rw http.ResponseWriter, r *http.Request) {
+	if !api.scimVerifyAuthHeader(r) {
+		_ = handlerutil.WriteError(rw, spec.Error{Status: http.StatusUnauthorized, Type: "invalidAuthorization"})
+		return
+	}
+
+	_ = handlerutil.WriteError(rw, spec.ErrNotFound)
+}
+
+// We currently use our own struct instead of using the SCIM package. This was
+// done mostly because the SCIM package was almost impossible to use. We only
+// need these fields, so it was much simpler to use our own struct. This was
+// tested only with Okta.
+type ScimUser struct {
+	Schemas  []string `json:"schemas"`
+	ID       string   `json:"id"`
+	UserName string   `json:"userName"`
+	Name     struct {
+		GivenName  string `json:"givenName"`
+		FamilyName string `json:"familyName"`
+	} `json:"name"`
+	Emails []struct {
+		Primary bool   `json:"primary"`
+		Value   string `json:"value"`
+		Type    string `json:"type"`
+		Display string `json:"display"`
+	} `json:"emails"`
+	Active bool          `json:"active"`
+	Groups []interface{} `json:"groups"`
+	Meta   struct {
+		ResourceType string `json:"resourceType"`
+	} `json:"meta"`
+}
+
+// scimPostUser creates a new user, or returns the existing user if it exists.
+func (api *API) scimPostUser(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	if !api.scimVerifyAuthHeader(r) {
+		_ = handlerutil.WriteError(rw, spec.Error{Status: http.StatusUnauthorized, Type: "invalidAuthorization"})
+		return
+	}
+
+	var sUser ScimUser
+	err := json.NewDecoder(r.Body).Decode(&sUser)
+	if err != nil {
+		_ = handlerutil.WriteError(rw, err)
+		return
+	}
+
+	email := ""
+	for _, e := range sUser.Emails {
+		if e.Primary {
+			email = e.Value
+			break
+		}
+	}
+
+	if email == "" {
+		_ = handlerutil.WriteError(rw, spec.Error{Status: http.StatusBadRequest, Type: "invalidEmail"})
+		return
+	}
+
+	user, _, err := api.AGPL.CreateUser(ctx, api.Database, agpl.CreateUserRequest{
+		CreateUserRequest: codersdk.CreateUserRequest{
+			Username: sUser.UserName,
+			Email:    email,
+		},
+		LoginType: database.LoginTypeOIDC,
```

Comment:
Should this be explicitly noted in the flag help? Not sure it's obvious that provisioned users would be OIDC.